### PR TITLE
Fix jagged composite impl of flatten()

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115842
* #116111
* __->__ #115192

Need to handle this in `NestedTensor.__torch_function__()` since it's CompositeImplicit

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]
Same comment about test, but otherwise stamped.

File: torch/nested/_internal/ops.py [hunk 1]
```diff
@@ -300,14 +301,24 @@ def _flatten_sig(input, start_dim=0, end_dim=-1):
...
         )
 
-        return NestedTensor(func(inp._values, **new_kwargs), **extract_kwargs(inp))
+        if start_dim == end_dim:
+            return inp
+
+        product = functools.reduce(operator.mul, inp.shape[start_dim : end_dim + 1])
+        new_shape = (
+            list(inp.shape[:start_dim]) + [product] + list(inp.shape[end_dim + 1 :])
+        )
```
Comment (soulitzer) [inline torch/nested/_internal/ops.py:319]: nit: `new_shape(*inp.shape[:start_dim], product, *inp.shape[end_dim + 1:])`

**Review by soulitzer — APPROVED** [review]

File: torch/nested/_internal/ops.py (hunk 1 above)
Comment (jbschlosser) [reply torch/nested/_internal/ops.py:319]: much nicer :0

**Review by soulitzer — APPROVED** [review]

**Pull request closed without merging**